state/pruner.go

```diff
@@ -0,0 +1,274 @@
+//
+// If a higher retain height is already set, we cannot accept the requested
+// height because the blocks might have been pruned.
+//
+// If the application has already set a retain height to a higher value we also
+// cannot accept the requested height as the blocks might have been pruned.
+func (p *Pruner) SetCompanionRetainHeight(height int64) error {
+	if height <= 0 || height < p.bs.Base() || height > p.bs.Height() {
+		return ErrInvalidHeightValue
+	}
+	currentCompanionRetainHeight, err := p.stateStore.GetCompanionBlockRetainHeight()
+	if err != nil {
+		if err == ErrKeyNotFound {
+			currentCompanionRetainHeight = height
+		} else {
+			return err
+		}
+	}
+	currentAppRetainHeight, err := p.stateStore.GetApplicationRetainHeight()
+	var noAppRetainHeight bool
+	if err != nil {
+		if err == ErrKeyNotFound {
+			noAppRetainHeight = true
+		} else {
+			return err
+		}
+	}
+	if currentCompanionRetainHeight > height || (!noAppRetainHeight && currentAppRetainHeight > height) {
+		return errors.New("cannot set a height lower than previously requested - blocks might have already been pruned")
+	}
+	err = p.stateStore.SaveCompanionBlockRetainHeight(height)
+	return err
+}
+
+// SetABCIResRetainHeight sets the retain height for ABCI responses.
+//
+// If the application has set the DiscardABCIResponses flag to true, nothing
+// will be pruned.
+func (p *Pruner) SetABCIResRetainHeight(height int64) error {
+	if height <= 0 || height > p.bs.Height() {
+		return ErrInvalidHeightValue
+	}
+	currentRetainHeight, err := p.stateStore.GetABCIResRetainHeight()
+	if err != nil {
+		if err == ErrKeyNotFound {
+			err = p.stateStore.SaveABCIResRetainHeight(height)
+			return err
+		}
+		return err
+	}
+	if currentRetainHeight > height {
+		return errors.New("cannot set a height lower than previously requested - blocks might have already been pruned")
+	}
+	err = p.stateStore.SaveABCIResRetainHeight(height)
+	return err
+}
+
+func (p *Pruner) pruningRoutine() {
+	p.logger.Info("Pruner started", "interval", p.interval.String())
+	lastHeightPruned := int64(0)
+	lastABCIResPrunedHeight := int64(0)
+	for {
+		select {
+		case <-p.Quit():
+			return
+		default:
+			lastHeightPruned = p.pruneBlocksToRetainHeight(lastHeightPruned)
+			lastABCIResPrunedHeight = p.pruneABCIResToRetainHeight(lastABCIResPrunedHeight)
+			time.Sleep(p.interval)
+		}
+	}
+}
+
+func (p *Pruner) pruneBlocksToRetainHeight(lastHeightPruned int64) int64 {
+	retainHeight := p.findMinRetainHeight()
+	if retainHeight != lastHeightPruned {
+		pruned, evRetainHeight, err := p.pruneBlocks(retainHeight)
+		if err != nil {
+			p.logger.Error("Failed to prune blocks", "err", err, "blockRetainHeight", retainHeight)
+		} else if pruned > 0 {
+			p.logger.Info("Pruned blocks", "count", pruned, "evidenceRetainHeight", evRetainHeight)
+		}
+		// The last height up until which we pruned is the current lowest point of the block store
+		// indicated by Base()
+		lastHeightPruned = p.bs.Base()
```

Comment:
Shouldn't this be:

```suggestion
		lastHeightPruned = p.bs.Base() - 1
```

because the base height is the lowest height we _have_, not the height to which we've pruned? The height to which we've pruned (which I thought was what was being returned here) is just below the base height.

Reply:
Good point! You are totally right.